### PR TITLE
Adding some extra shared lanes to iOS Fastfile to allow for more dynamic code signing and optional OneSky downloading

### PR DESCRIPTION
**Pull request by levieggertcru**
Took existing logic in the iOS Fastfile and pulled things out into separate lanes.  We will also need to change ```automatic_code_signing``` in the iOS Fastfile to use ```update_code_signing_settings``` due to deprecation. 

- ```cru_shared_lane_run_tests``` A lane for running tests against provided scheme.
- ```cru_shared_lane_download_and_commit_latest_one_sky_localizations``` A separate lane for downloading OneSky translations and committing changes to.  Opted for a separate lane for this because it's possible not all iOS apps will use OneSky. The current OneSky download is also using an ENV["CRU_SCHEME"] which doesn't make any sense because OneSky downloads have nothing to do with an Xcode Scheme.  That is actually a directory name where localizations are downloaded to.
- ```cru_shared_lane_increment_xcode_project_build_number``` A lane for incrementing the xcode project build number based on the latest TestFlight release build number.  Requires api_key_path.
- ```cru_shared_lane_build_and_deploy_for_testflight_release``` A lane that will configure Xcode code signing, run match, and build an app for release distributed to TestFlight.  In this I also included the argument ```skip_waiting_for_build_processing``` set to value ```true``` which will cut down about 20-25 minutes on CI time.  User's are still notified via email and notifications if they have notifications enabled for the TestFlight app.

### CODE REVIEW

File: Fastfile [hunk 1]
```diff
@@ -295,4 +295,231 @@ platform :ios do
...
+  #
+  # options:
+  # - podfile: Explicitly specify the path to the Cocoapods' Podfile. You can either set it to the Podfile's path or to the folder containing the Podfile file.
+  #
+  lane :cru_shared_lane_pod_install do |options|
+
+    podfile = options[:podfile] || "./Podfile"
+
+    cocoapods(podfile: podfile)
+  end
```
Comment (frett) [inline Fastfile:313]: this can be removed since you have already switched things back to using `pod install` directly

File: Fastfile [hunk 2]
```diff
@@ -295,4 +295,231 @@ platform :ios do
...
+  # - one_sky_secret_key:
+  #
+  lane :cru_shared_lane_download_and_commit_latest_one_sky_localizations do |options|
+
+    one_sky_download_to_project_directory = options[:one_sky_download_to_project_directory] || ENV["ONESKY_DOWNLOAD_TO_XCODE_PROJECT_DIRECTORY"]
+    one_sky_filename = options[:one_sky_filename] || ENV["ONESKY_FILENAME"]
+    one_sky_localizations_string = options[:one_sky_localizations] || ENV["ONESKY_LOCALIZATIONS"]
+    one_sky_project_id = options[:one_sky_project_id] || ENV["ONESKY_PROJECT_ID"]
+    one_sky_public_key = options[:one_sky_public_key] || ENV["ONESKY_PUBLIC_KEY"]
+    one_sky_secret_key = options[:one_sky_secret_key] || ENV["ONESKY_SECRET_KEY"]
```
Comment (frett) [inline Fastfile:350]: I don't think any of our apps would actually require the secret key, but I'm fine with leaving this here anyways

File: Fastfile [hunk 3]
```diff
@@ -295,4 +295,231 @@ platform :ios do
...
+  #
+  # options:
+  # - api_key_path: Path to your App Store Connect API Key JSON file.
+  # - apple_id_username: Your Apple ID Username. 
+  # - app_release_bundle_identifier: The bundle identifier for your xcode release configuration.
+  #
+  lane :cru_shared_lane_increment_xcode_project_build_number do |options|
+
+    api_key_path = options[:api_key_path] || ENV["APP_STORE_CONNECT_API_KEY_JSON_FILE_PATH"]
+    apple_id_username = options[:apple_id_username] || ENV["APPLE_ID_USERNAME"]
```
Comment (frett) [inline Fastfile:398]: is `apple_id_username` still required? I believe switching to the `api_key_path` removes the need for a username

File: Fastfile [hunk 4]
```diff
@@ -295,4 +295,231 @@ platform :ios do
...
+    app_release_bundle_identifier = options[:app_release_bundle_identifier] || ENV["APP_RELEASE_BUNDLE_IDENTIFIER"]
+
+    latest_build_number = latest_testflight_build_number(
+        api_key_path: api_key_path,
+        app_identifier: app_release_bundle_identifier,
+        username: apple_id_username
```
Comment (frett) [inline Fastfile:404]: see previous comment about this potentially not being needed

File: Fastfile [hunk 5]
```diff
@@ -295,4 +295,231 @@ platform :ios do
...
+  # - match_git_url:
+  # - match_keychain_name:
+  # - team_id: The ID of your App Store Connect team.
+  #
+  lane :cru_shared_lane_build_and_deploy_for_testflight_release do |options|
+
+    # Update code signing settings
+
+    api_key_path = options[:api_key_path] || ENV["APP_STORE_CONNECT_API_KEY_JSON_FILE_PATH"]
+    apple_id_username = options[:apple_id_username] || ENV["APPLE_ID_USERNAME"]
```
Comment (frett) [inline Fastfile:440]: `apple_id_username` may not be needed

File: Fastfile [hunk 6]
```diff
@@ -295,4 +295,231 @@ platform :ios do
...
+        app_identifier: app_bundle_ids_array,
+        git_basic_authorization: Base64.strict_encode64(ENV["MATCH_GIT_BASIC_AUTHORIZATION_PAT"]),
+        git_branch: match_git_branch,
+        git_url: match_git_url,
+        keychain_name: match_keychain_name,
+        keychain_password: ENV["MATCH_PASSWORD"],
+        platform: "ios",
+        storage_mode: "git",
+        type: "appstore",
+        username: apple_id_username
```
Comment (frett) [inline Fastfile:500]: `username` may not be necessary anymore

File: Fastfile [hunk 7]
```diff
@@ -295,4 +295,231 @@ platform :ios do
...
+    )
+
+    # TestFlight - Release
+
+    testflight(
+        app_identifier: app_release_bundle_identifier,
+        dev_portal_team_id: team_id,
+        ipa: release_ipa_path,
+        skip_waiting_for_build_processing: true,
+        username: apple_id_username
```
Comment (frett) [inline Fastfile:522]: another `username` that may not be necessary since it should use the `api_key` now

File: Fastfile (hunk 2 above)
Comment (levieggertcru) [reply Fastfile:350]: Hey @frett that's correct MPDX and GodTools aren't requiring it, but would be good to keep it.

File: Fastfile (hunk 3 above)
Comment (levieggertcru) [reply Fastfile:398]: Hey @frett let me check and see if we can drop this.

Comment (frett) [reply Fastfile:384]: I know Apple started requiring MFA for accounts at the beginning of 2021, at which time we transitioned to using the API key. prior to that it was username & password to authenticate. The username field is leftover from prior to that change.

**Comment by frett** [comment]
interesting, didn't realize `team_id` was no longer necessary either, guessing it makes sense because that is bundled in the API key json

File: Fastfile (hunk 3 above)
Comment (levieggertcru) [reply Fastfile]: @frett I removed those, I think you're right.

**Review by frett — APPROVED** [review]

**Pull request merged**